patches/server/0949-Paper-Plugins.patch

```diff
@@ -0,0 +1,6590 @@
++                unblock(w);
++            }
++        }
++    }
++
++    @SuppressWarnings("unchecked")
++    private void initState(Consumer<List<V>> consumer)
++    {
++        cycleConsumer = consumer;
++        iToV = (V[]) graph.nodes().toArray();
++        vToI = new HashMap<>();
++        blocked = new HashSet<>();
++        bSets = new HashMap<>();
++        stack = new ArrayDeque<>();
++
++        for (int i = 0; i < iToV.length; i++) {
++            vToI.put(iToV[i], i);
++        }
++    }
++
++    private void clearState()
++    {
++        cycleConsumer = null;
++        iToV = null;
++        vToI = null;
++        blocked = null;
++        bSets = null;
++        stack = null;
++    }
++
++    private void initMinSCGState()
++    {
++        index = 0;
++        foundSCCs = new ArrayList<>();
++        vIndex = new HashMap<>();
++        vLowlink = new HashMap<>();
++        path = new ArrayDeque<>();
++        pathSet = new HashSet<>();
++    }
++
++    private void clearMinSCCState()
++    {
++        index = 0;
++        foundSCCs = null;
++        vIndex = null;
++        vLowlink = null;
++        path = null;
++        pathSet = null;
++    }
++
++    private Integer toI(V vertex)
++    {
++        return vToI.get(vertex);
++    }
++
++    private V toV(Integer i)
++    {
++        return iToV[i];
++    }
++
++    private Set<V> getBSet(V v)
++    {
++        // B sets typically not all needed,
++        // so instantiate lazily.
++        return bSets.computeIfAbsent(v, k -> new HashSet<>());
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/LegacyPluginLoadingStrategy.java b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/LegacyPluginLoadingStrategy.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..fc80c7f1746a5f62b3f76646db4b91b5caca6655
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/LegacyPluginLoadingStrategy.java
+@@ -0,0 +1,253 @@
++package io.papermc.paper.plugin.entrypoint.strategy;
++
++import com.google.common.graph.GraphBuilder;
++import com.google.common.graph.MutableGraph;
++import io.papermc.paper.plugin.entrypoint.dependency.GraphDependencyContext;
++import io.papermc.paper.plugin.configuration.PluginMeta;
++import io.papermc.paper.plugin.provider.PluginProvider;
++import org.bukkit.plugin.UnknownDependencyException;
++
++import java.util.*;
++import java.util.logging.Level;
++import java.util.logging.Logger;
++
++@SuppressWarnings("UnstableApiUsage")
++public class LegacyPluginLoadingStrategy<T> implements ProviderLoadingStrategy<T> {
++
++    private static final Logger LOGGER = Logger.getLogger("LegacyPluginLoadingStrategy");
++
++    private final ProviderConfiguration<T> configuration;
++
++    public LegacyPluginLoadingStrategy(ProviderConfiguration<T> onLoad) {
++        this.configuration = onLoad;
++    }
++
++    @Override
++    public List<T> loadProviders(List<PluginProvider<T>> providers) {
```

Comment:
The class you mentioned this in is the legacy spigot plugin loader. This diff is basically just the normal spigot loader but moved over.

Are we sure that we want to split the spigot logic up? I kept this identical in order to easily port changes that spigot might make.

Reply:
It's already modified by us, and since we'll have to manually copy them over anyways, just moving individual parts into extra methods won't make it meaningfully more difficult

Reply:
![image](https://user-images.githubusercontent.com/23108066/214958665-6ec33fdf-af06-47d0-9f50-efab138217ca.png)
In general it's kind of messy, it would just be a bunch of methods with a ton of parameters. Is it still really worth splitting up this behavior? 

The spigot loader is very confusing, like, there are *two* places where plugin loading may actually occur. 
![image](https://user-images.githubusercontent.com/23108066/214958918-1807e0d0-f246-447a-84c6-aeebe29a932e.png)
![image](https://user-images.githubusercontent.com/23108066/214958939-f7ae77d2-7ca8-4c73-939b-383b39f4c6bd.png)

Reply:
That can be cleaned up later then